src/state/machine.hpp

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include "state/events.hpp"
+#include "state/listening.hpp"
+#include "state/sleeping.hpp"
+#include "state/state.hpp"
+#include <typeinfo>
+
+namespace genie {
+
+class App;
+
+namespace state {
+
+// State Machine
+// ===========================================================================
+
+struct GMainLoopEvent {
+  Machine *machine;
+  events::Event *event;
+};
+
+class Machine {
+public:
+  App *app;
+  State *current_state;
+
+  Machine(App *app) : app(app){};
+
+  template <typename S> void init() {
+    this->current_state = new S(this);
+    this->current_state->enter();
+  }
+
+  template <typename E> guint dispatch(E *event) {
+    // g_message("DISPATCH EVENT %s", typeid(E).name());
+    GMainLoopEvent *loop_event = new GMainLoopEvent();
+    loop_event->machine = this;
+    loop_event->event = event;
+    return g_idle_add(handle<E>, loop_event);
+  }
+
+  template <typename E> static gboolean handle(gpointer user_data) {
+    // g_message("HANDLE EVENT %s", typeid(E).name());
+    GMainLoopEvent *loop_event = static_cast<GMainLoopEvent *>(user_data);
+    E *event = static_cast<E *>(loop_event->event);
+    loop_event->machine->current_state->react(event);
+    delete event;
```

Comment:
`delete loop_event` here too

Reply:
Oh even shit I noticed I needed that but then never added it. I think I planned to have the `GMainLoopEvent` delete the event on destruction and then just delete the `loop_event` but forgot about it

Reply:
Fixed in e22c9a940416d2d32c8ca73a241603c65fd31d45